### PR TITLE
[WIP] Make sure NotSelector selects series with label not set

### LINKED ISSUE #3575: Change in selector behaviour for !=
**What did you do?**
query:
`metrics{label != "thing"}` 
where label is not set.
 
**What did you expect to see?**
In prometheus 1.6.3 this will return all the items where label != "thing", including items where label is not set at all.

**What did you see instead? Under which circumstances?**
In prometheus 2.0 items without label set are not returned.

### PR DESCRIPTION
**Pull request by Gouthamve**
Fixes #3575

Comes with a slight performance penalty which is better off addressed in tsdb

Tests pending but would like an preliminary look to see if approach is right @fabxc @brian-brazil

---
This change is [Reviewable](https://reviewable.io/reviews/prometheus/prometheus/3578)

### CODE REVIEW

File: storage/tsdb/tsdb.go [hunk 1]
```diff
@@ -190,14 +190,39 @@ type querier struct {
...
+	var err error
+	notSeen := false
 
-	for _, om := range oms {
+	for i, om := range oms {
 		ms = append(ms, convertMatcher(om))
+
+		// See: https://github.com/prometheus/prometheus/issues/3575
+		if om.Type == labels.MatchNotRegexp || om.Type == labels.MatchNotEqual {
+			if notSeen == false {
```
Comment (brian-brazil) [inline storage/tsdb/tsdb.go]: This doesn't seem right, there can be multiple negative matchers.

Comment (Gouthamve) [reply storage/tsdb/tsdb.go]: This is right because there is a recursive call that handles other matchers.

**Comment by fabxc** [comment]
While it is not a lot of code here, I find it extremely hard to grasp.
This indeed seems like something where we should figure out our defined semantics in TSDB and ideally handle it there.

Do I see it correctly that one problem of it doing it directly there is that matchers are an interface and it's not as easy to infer things? We are doing [type casting](https://github.com/prometheus/tsdb/blob/master/querier.go#L213) there already for some cases, so doing it for this here as well wouldn't add massive craziness.

We can probably address this issue here but also our rather in-elegant use of a list of `absent` labels to check after retrieving (potentially a lot of) superfluous series.

Would the following work: In the tsdb/queriery.go code linked above, do not cast to `EqualMatcher` but generally check whether `m.Matchers("")`. This then also includes `!="something"` or `!~"something"`. If we encounter this, get all postings list for label values `v` where `!m.Matches(v)`, create a merge iterator over them, and wrap this one in an inversion iterator (returns everything not found in the merged set) and intersect this with the postings list containing all series.

I think the intersecting against the list of all postings is not that expensive. But merging a lot of those other small postings certainly can be. We could add another postings list `lname=<has_value>` that contains all IDs for a label name that has have some value `!=""` set.

**Comment by Gouthamve** [comment]
Yes, what I don't like is that the detail that `!=` selects series with the label unset is also flowing down into tsdb. It's un-intuitive and I am not sure tsdb should be handling it that way.

Will whip up the PR with the fix tonight anyways though.

**Comment by fabxc** [comment]
PR in TSDB exists now but could you elaborate why you find in un-intuitive?

**Comment by Gouthamve** [comment]
Superseded by: https://github.com/prometheus/tsdb/pull/228

**Pull request closed without merging**